### PR TITLE
Build DOM via JSX

### LINKED ISSUE #270: Audit extension for XSS vulns
https://github.com/sindresorhus/refined-github/pull/269#issuecomment-233547355

We likely have more than just that 1. Note that the risk factor is lower than usual thanks to GitHub using a strict content security policy, but it's still worthwhile.

### LINKED ISSUE #539: Using JSX to create elements XSS-free
We use HTML strings with user-provided data all over the extension, this is XSS 101.

The alternative to that is a lot of inelegant calls and de-XSS-ifying [this piece of code](https://github.com/sindresorhus/refined-github/blob/82c11c1a1906bc735892256f25295e436153b53a/src/libs/mark-unread.js#L187-L214) gets ugly real fast.

I wish there was a JSX-to-DOM compiler that allowed use to write safe HTML:

```jsx
const element = <a class="one two" href="{element.title}">{name}</a>
```
```js
const element = document.createElement('a');
element.className = 'one two';
element.href = element.title;
element.textContent = name;
```

Well, it turns out we have some options:

https://www.npmjs.com/package/rollup-plugin-jsx

A factory for that would be something like [`h`](https://github.com/dominictarr/h/blob/master/index.js), which appears to be pretty small.

Just like that, all `.innerHTML` calls are gone.

### PR DESCRIPTION
**Pull request by vadimdemedes**
Fixes #539 (JSX)
Fixes #270 (XSS audit)

Uses [dom-chef](https://github.com/vadimdemedes/dom-chef) to build DOM elements via JSX on the fly.

All the other pages except notifications are broken, because icons (`src/libs/icons.js`) are now DOM elements, not strings. I thought it would be fine for now to demonstrate this PR.

Let me know what you think ;)

### CODE REVIEW

**Comment by bfred-it** [comment]
Nice! But [h]() is much lighter and it has no dependencies. https://github.com/dominictarr/h/blob/master/index.js

**Comment by vadimdemedes** [comment]
Oh, when you said:

> A factory for that would be something like h, which appears to be pretty small.

I thought `h` was just an example, not the actual module you wanted to use.

In `dom-chef`'s defence, here's what `dom-chef` has and `h` doesn't:

- SVG support (all the icons in refined-github)
- Full test coverage
- Support. Looks like `h` is abandoned a while ago in favor of `hyperscript`, see https://github.com/dominictarr/h/issues/1#issuecomment-161162232.

**Comment by bfred-it** [comment]
Alright. Can you fix the issues? `xo --fix` should do most of them.

**Comment by bfred-it** [comment]
I tried this and it works! 🎉  It's great to see HTML's syntax highlighting rather than the previous JS strings.

What else do you think this is missing besides fixing the icons? If this is good for you I can pick it up from here and fix the icons. That'll require the JSX-ification of most of the components the use the icons.

For anyone wondering, this is what the generated code looks like:

```js
const item = (
	h_1('li', {class: `list-group-item js-notification js-navigation-item unread ${type}-notification`}, [
		h_1('span', {class: "list-group-item-name css-truncate"}, [
			icon,

			h_1('a', {href: url, class: "css-truncate-target js-notification-target js-navigation-open list-group-item-link"}, [
				title
			])
		]),

		h_1('ul', {class: "notification-actions"}, [
			h_1('li', {class: "delete"}, [
				h_1('button', {'aria-label': "Mark as read", class: "btn-link delete-note tooltipped tooltipped-w js-mark-read"}, [
					check
				])
			]),

			h_1('li', {class: "mute"}, [
				h_1('button', {style: {opacity: 0, pointerEvents: 'none'}}, [
					mute
				])
			]),

			h_1('li', {class: "age"}, [
				h_1('relative-time', {datetime: date, title: dateTitle})
			]),

			h_1('li', {class: "tooltipped tooltipped-s", 'aria-label': usernames}, [
				h_1('div', {class: "avatar-stack clearfix"}, [
					avatars
				])
			])
		])
	])
```

**Comment by vadimdemedes** [comment]
> What else do you think this is missing besides fixing the icons?

Aside from icons, which are imported across all modules, I don't think there's anything else.

> If this is good for you I can pick it up from here and fix them.

Sure, feel free :)

**Comment by vadimdemedes** [comment]
I'm wondering though why it's `h_1` and not `h`?

**Comment by bfred-it** [comment]
Still a WIP, 50% tested. Waiting for some PRs to be merged so I can JSX-ify their template strings as well.

**Comment by bfred-it** [comment]
WIP. Waiting for some PRs to be merged so I can JSX-ify their template strings as well.

Tested each piece separately, except `addDeleteForkLink` because I don't see it :D 

Essentially the only annoyance I found with JSX is that babel discards the whitespace between elements, which is sometimes meaningful (only twice in the entire extension, so it's not that annoying)

**Comment by bfred-it** [comment]
I decided that this PR doesn't need **everything** to be ready but it's good enough as is:

- adds JSX to the build
- JSX-ifies `mark-unread` as proof of concept

This build and `mark-unread` have been tested.

**Comment by bfred-it** [comment]
Ok, let's get this all in. The rest will come later.

File: src/content.js [hunk 1]
```diff
@@ -254,14 +254,12 @@ function addPatchDiffLinks() {
...
 
-	const $commitMeta = $('.commit-meta span.float-right');
-
-	$commitMeta.append(`
+	select('.commit-meta span.float-right').append(
 		<span class="sha-block patch-diff-links">
-			<a href="${commitUrl}.patch" class="sha">patch</a>
-			<a href="${commitUrl}.diff" class="sha">diff</a>
+			<a href={`${commitUrl}.patch`} class="sha">patch</a>
+			<a href={`${commitUrl}.diff`} class="sha">diff</a>
```
Comment (sindresorhus) [inline src/content.js]: The space between these links are now lost.

File: src/content.js [hunk 2]
```diff
@@ -406,14 +406,14 @@ function addOPLabels() {
 }
 
 function addMilestoneNavigation() {
-	$('.repository-content').before(`
+	select('.repository-content').insertAdjacentElement('beforeBegin',
```
Comment (sindresorhus) [inline src/content.js]: beforeBegin => beforebegin

File: src/libs/reactions-avatars.js [hunk 3]
```diff
@@ -29,17 +31,20 @@ function add(currentUser) {
...
 			const firstThreeParticipants = participants.slice(0, 3);
-			const $participantsContainer = $element.find('.participants-container');
+			const participantsContainer = element.querySelector('.participants-container');
 
 			// Clear any existing avatars and remainder count
-			$participantsContainer.html('');
+			// https://stackoverflow.com/a/3955238/288906
+			while (participantsContainer.firstChild) {
+				participantsContainer.firstChild.remove();
+			}
```
Comment (sindresorhus) [inline src/libs/reactions-avatars.js]: This should be moved to utils.

**Comment by sindresorhus** [comment]
The `import {h} from 'dom-chef';` in every file is kinda annoying.

---

https://github.com/sindresorhus/refined-github/pull/558/commits/084bbd9026b5e6e5bc0a0b6f00207f516b1bdd27

> Remove automatic dom-chef importer. Unfortunately babel-plugin-jsx-pragmatic causes issues at runtime.

@bfred-it What issues?

---

@vadimdemedes Could you maybe make a plugin like https://github.com/jmm/babel-plugin-jsx-pragmatic but specifically for `dom-chef`, so it would just be plug-and-play?

**Comment by bfred-it** [comment]
> > Remove automatic dom-chef importer. Unfortunately babel-plugin-jsx-pragmatic causes issues at runtime.
>
> [@bfred-it](https://github.com/bfred-it) What issues?

The generated code has runtime errors. I felt it was unnecessary to add/debug a build step to skip a line of code.

error


> Could you maybe make a plugin like [jmm/babel-plugin-jsx-pragmatic](https://github.com/jmm/babel-plugin-jsx-pragmatic) but specifically for `dom-chef`

Why? Why should it be specific?

**Comment by sindresorhus** [comment]
> The generated code has runtime errors. I felt it was unnecessary to add/debug a build step to skip a line of code.

I see. We don't have to fix it for this PR to land. Was just suggestion some future improvements.

> Why? Why should it be specific?

To simplify the config and not having to specify a custom pragma, but if we do a `dom-chef` Babel preset, I guess the plugin could be un-opinionated.

**Comment by sindresorhus** [comment]
This PR is ready to land as soon as https://github.com/sindresorhus/refined-github/pull/558#discussion_r126108501 and https://github.com/sindresorhus/refined-github/pull/558#discussion_r126109133 are resolved.

File: src/content.js (hunk 1 above)
Comment (bfred-it) [inline src/content.js]: Ah yeah, that's the *annoyance* I mentioned in https://github.com/sindresorhus/refined-github/pull/558#issuecomment-313317987

Fixing

**Comment by sindresorhus** [comment]
I opened an issue about the Babel preset stuff: https://github.com/vadimdemedes/dom-chef/issues/7

File: src/content.js (hunk 1 above)
Comment (bfred-it) [inline src/content.js]: Fixed

Comment (sindresorhus) [inline src/content.js]: @vadimdemedes Is this expected JSX behavior or should we open an issue somewhere?

**Comment by sindresorhus** [comment]
Sidenote: This PR brings to build time from 571ms to 1703ms.

**Comment by bfred-it** [comment]
Interesting. I just noticed Travis actually runs on the merged version of the repo, not just the branch: https://github.com/sindresorhus/refined-github/commit/b06f0b46820eba1919836712d5252b0f893d7193

**Comment by sindresorhus** [comment]
Landed it manually so I could preserve the credit for both of you :)

**Comment by bfred-it** [comment]
Thanks! I was building onto that PR thinking that the merge commit would include both names when I rebased it :D

**Comment by bfred-it** [comment]
@sindresorhus build time is 700ms with `nativejsx-loader`. Replacement is easy:

```diff
-					loader: 'babel-loader'
+					loader: 'nativejsx-loader'
```

But it fails with async/await: https://github.com/vadimdemedes/dom-chef/issues/7#issuecomment-313681873

**Pull request closed without merging**